### PR TITLE
✨ Add DangerousWorkflow check for imposter commits.

### LINKED ISSUE #2733: Feature: Dangerous Workflows - Imposter Commit Check
**Is your feature request related to a problem? Please describe.**

We disclosed an issue with GitHub Actions today that lets users bypass workflow settings by using forked commits - https://www.chainguard.dev/unchained/what-the-fork-imposter-commits-in-github-actions-and-ci-cd

I'd like to contribute a check to Scorecards based on our proof-of-concept tool - https://github.com/chainguard-dev/clank

**Describe the solution you'd like**

I'd like to add an additional check to `Dangerous-Workflows` based on https://github.com/chainguard-dev/clank (this check is most similar to the `Untrusted Code Checkout` check).

It roughly works by inspecting each `uses` statement then seeing if the commit is reachable by a branch or tag in the repo.

**Describe alternatives you've considered**

- An early version of clank cloned the repo to a local tmp to avoid extra network calls. I ended up having some difficulty trying to handle untracked branches, but we can revisit this if the reachability queries end up being too costly.
- Tried experimenting with the GraphQL API to detect reachability, but had some difficulty forming a query that wasn't a form of "GraphQL get every reachable commit SHA then check if it's in the list". See https://github.com/orgs/community/discussions/24501 for related discussion. If anyone has ideas for how to get GraphQL to work in a reasonable way, lmk!

**Additional context**

I'm happy to contribute this!

WIP branch already started here - https://github.com/ossf/scorecard/compare/main...wlynch:scorecard:imposter-commits
Functionally, but need to add tests. Open to early feedback! 🙏 

If anyone has any suggestions/examples for how to create sub-clients of a githubclient RepoClient, let me know. Wasn't sure how to tackle this for the mock clients in the tests, and I don't want to start refactoring clients without consulting maintainers first.

### PR DESCRIPTION
**Pull request by wlynch**
#### What kind of change does this PR introduce?

Adds DangerousWorkflow check for imposter commits.
    
See https://www.chainguard.dev/unchained/what-the-fork-imposter-commits-in-github-actions-and-ci-cd

This borrows its implementation from
https://github.com/chainguard-dev/clank to look up imposter commits for
a repo.

(Is it a bug fix, feature, docs update, something else?)

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

n/a (new check)

#### What is the new behavior (if this is a feature change)?**

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Fixes #2733 

#### Special notes for your reviewer

GitHub e2e test is failing because the Actions yaml is testing against isn't actually valid (it references non-existent repos). We should fix this, but figure get this PR out for review.

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note
Adds new DangerousWorkflow check for detecting imposter commits.
```

### CODE REVIEW

File: clients/gitlabrepo/client.go [hunk 1]
```diff
@@ -135,6 +135,36 @@ func (client *Client) InitRepo(inputRepo clients.Repo, commitSHA string, commitD
...
+		searchCommits: client.searchCommits,
+		webhook:       client.webhook,
+		languages:     client.languages,
+		licenses:      client.licenses,
+	}
+	repo, err := MakeGitlabRepo(inputRepo)
+	if err != nil {
+		return nil, err
+	}
+	if err := newClient.InitRepo(repo, commitSHA, commitDepth); err != nil {
```
Comment (spencerschrock) [inline clients/gitlabrepo/client.go:162]: similar concerns

**Review by spencerschrock — CHANGES_REQUESTED** [review]
Not a complete review, but wanted to get some comments out that address some big blockers with this current implementation. This change will likely require converting `Dangerous-Workflow` away from a file content only check:
https://github.com/ossf/scorecard/blob/eb57e04ac487a76f4dd71aa3cdfc2806c5cd5076/checks/dangerous_workflow.go#L28-L32

I'll also post some thoughts in the main issue for this PR that are better suited there.

**Comment by spencerschrock** [comment]
I agree with the need for a sub-`RepoClient`, as we're trying to answer the question "does this commit belong to repo B" but I'm not sure how I feel about the current approach. A new method like `NewClient` makes sense since `RepoClient` might be a github/gitlab/local etc, but I'm wondering if there's a way we can make the child `RepoClient` lighter. 

If we modify RepoClient so the struct fields are optional that could help:

e.g.
```
if client.graphClient != nil {
    client.graphClient.init(client.ctx, client.repourl, client.commitDepth)
}
```

In the GitHub implementation for example,  the `ContainsRevision` method only uses the `branchesHandler` handler. It does it to read `repourl` which is already a member of `Client`, so we could even eliminate the need to use `branchesHandler`.

**Comment by spencerschrock** [comment]
I think it makes sense to discuss this some more before re-implementing things.

**Comment by wlynch** [comment]
For the repo client - good to know that the handlers are stateful. I think we could probably get a fresh client if we just refactor some of the client initialization from here - https://github.com/ossf/scorecard/blob/b6362b14ee0bc2dd2d6d3fbc7ccff7fbd7cd57ae/clients/githubrepo/client.go#L265-L309

The main piece we want to reuse is the authed client, which we can pick out from the client struct. I'll take a pass at it. (also open to other ideas)

File: clients/githubrepo/client.go [hunk 2]
```diff
@@ -124,6 +124,35 @@ func (client *Client) InitRepo(inputRepo clients.Repo, commitSHA string, commitD
...
+		webhook:       client.webhook,
+		languages:     client.languages,
+		licenses:      client.licenses,
+		tarball:       client.tarball,
+	}
+	repo, err := MakeGithubRepo(inputRepo)
+	if err != nil {
+		return nil, err
+	}
+	if err := newClient.InitRepo(repo, commitSHA, commitDepth); err != nil {
```
[the remark this answers is not in the document]
Comment (wlynch) [reply clients/githubrepo/client.go]: Refactored this to use the same handler creation logic as CreateGithubRepoClientWithTransport. PTAL!

**Comment by wlynch** [comment]
> This change will likely require converting Dangerous-Workflow away from a file content only check:

Looks like CommitBased / FileBased are the only 2 types at the moment. https://github.com/ossf/scorecard/blob/eb57e04ac487a76f4dd71aa3cdfc2806c5cd5076/checker/check_request.go#L40-L45

I assume we'll need to make a new one? Any suggestions for what this new type should be? `APIBased`?

**Comment by spencerschrock** [comment]
> > This change will likely require converting Dangerous-Workflow away from a file content only check:
> 
> Looks like CommitBased / FileBased are the only 2 types at the moment.
> 
> https://github.com/ossf/scorecard/blob/eb57e04ac487a76f4dd71aa3cdfc2806c5cd5076/checker/check_request.go#L40-L45
> 
> I assume we'll need to make a new one? Any suggestions for what this new type should be? `APIBased`?

I meant more along the lines of removing `FileBased` from the `Dangerous-Workflow` check as it now requires the API call. I don't think we need an `APIBased` as every check is assumed to be API based unless FileBased is declared.

https://github.com/ossf/scorecard/blob/b6362b14ee0bc2dd2d6d3fbc7ccff7fbd7cd57ae/checks/dangerous_workflow.go#L28-L32

Which is a little unfortunate, as there's still plenty of functionality which works on the file content. Scorecard would benefit from some sort of "optional API call" as once a check removes support for `FileBased` it can't be run locally.

```
go run main.go --local=. --checks Dangerous-Workflow
Error: GetEnabled: internal error: Unsupported RequestType [0] by check: Dangerous-Workflow
2023/04/03 14:35:00 error during command execution: GetEnabled: internal error: Unsupported RequestType [0] by check: Dangerous-Workflow
```

**Comment by spencerschrock** [comment]
> For the repo client - good to know that the handlers are stateful. I think we could probably get a fresh client if we just refactor some of the client initialization from here -
> 
> https://github.com/ossf/scorecard/blob/b6362b14ee0bc2dd2d6d3fbc7ccff7fbd7cd57ae/clients/githubrepo/client.go#L265-L309
> 
> The main piece we want to reuse is the authed client, which we can pick out from the client struct. I'll take a pass at it. (also open to other ideas)

Agree that extracting the auth'd transport is the main bit we care about. I'll take a look at any follow-up commits later.

We can always revisit the optional `init()` comment if performance becomes a problem. The tarball handler's init function for example downloads the tarball and extracts it to a `/tmp` directory. So all of these subclients would be doing extra work when we really just care about is making an API call using the auth from the main client.

**Comment by spencerschrock** [comment]
Disregard about the tarball client, it's loaded lazily so shouldn't be a problem

**Review by spencerschrock — COMMENTED** [review]
The NewClient approach looks good to me. 

I've left some comments on the ref parsing, with regard to re-usable workflows (which should be vulnerable to imposter commits) as well as filtering out non commit SHA refs.

And finally, the increased quota would likely be a problem for running this check in the cron so want to get some input from @azeemshaikh38 on that one.

File: checks/raw/dangerous_workflow.go [hunk 3]
```diff
@@ -269,3 +282,102 @@ func checkVariablesInScript(script string, pos *actionlint.Pos,
...
+			if len(s) != 2 {
+				return sce.WithMessage(sce.ErrorCheckRuntime, fmt.Sprintf("unexpected reference: %s", trimmedRef))
+			}
+			// Repo references can include paths (e.g. github/codeql-action/init) - Trim first n
+			repoSplit := strings.SplitN(s[0], "/", 3)
+			if len(repoSplit) < 2 {
+				return sce.WithMessage(sce.ErrorCheckRuntime, fmt.Sprintf("unexpected repo reference: %s", s[0]))
+			}
+			repo := strings.Join(repoSplit[:2], "/")
+			sha := s[1]
```
[the remark this answers is not in the document]
Comment (spencerschrock) [reply checks/raw/dangerous_workflow.go:314]: would also need to force `sha` to lowercase, or account for it in the regex

File: checks/dangerous_workflow.go [hunk 4]
```diff
@@ -26,11 +26,10 @@ const CheckDangerousWorkflow = "Dangerous-Workflow"
 
 //nolint:gochecknoinits
 func init() {
-	supportedRequestTypes := []checker.RequestType{
-		checker.FileBased,
-		checker.CommitBased,
-	}
-	if err := registerCheck(CheckDangerousWorkflow, DangerousWorkflow, supportedRequestTypes); err != nil {
```
[the remark this answers is not in the document]
Comment (azeemshaikh38) [reply checks/dangerous_workflow.go:33]: +1

File: clients/repo_client.go [hunk 5]
```diff
@@ -30,6 +30,7 @@ const HeadSHA = "HEAD"
 // RepoClient interface is used by Scorecard checks to access a repo.
 type RepoClient interface {
 	InitRepo(repo Repo, commitSHA string, commitDepth int) error
+	NewClient(repo string, commitSHA string, commitDepth int) (RepoClient, error)
```
Comment (azeemshaikh38) [inline clients/repo_client.go:33]: I'm not sure I understand the usecase for `NewClient` API. Could we not use `CreateGitHubRepoClient` instead?

File: checks/raw/dangerous_workflow.go [hunk 6]
```diff
@@ -269,3 +282,102 @@ func checkVariablesInScript(script string, pos *actionlint.Pos,
...
+func validateImposterCommits(client clients.RepoClient, workflow *actionlint.Workflow, path string,
+	pdata *checker.DangerousWorkflowData,
+) error {
+	ctx := context.TODO()
+	cache := &containsCache{
+		client: client,
+		cache:  make(map[commitKey]bool),
+	}
+	for _, job := range workflow.Jobs {
+		for _, step := range job.Steps {
```
[the remark this answers is not in the document]
Comment (azeemshaikh38) [reply checks/raw/dangerous_workflow.go:295]: Reusable workflows can't be hash pinned. Would be interesting to check whether the imposter commit holds true for reusable workflows too.

**Review by azeemshaikh38 — COMMENTED** [review]
Some high-level comments about code architecture:

- We could avoid changes to the `RepoClient` interface altogether and instead inject a new dependency in [RunScorecard](https://github.com/ossf/scorecard/blob/main/pkg/scorecard.go#L84) fn.
- For the cron job, we can later optimize this injected dependency with a more API efficient model (like a blobstore-based cache maybe) while regular CLI continues to use REST API.
- Flag-guard this change so that it does not get rolled out to prod without some testing on our end.

File: checks/raw/dangerous_workflow.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (spencerschrock) [reply checks/raw/dangerous_workflow.go:295]: Taken from the docs:
> {ref} can be a SHA, a release tag, or a branch name.

And an experiment I did last year for a different reason
https://github.com/spencerschrock/reusable-workflow-caller/blob/aa0ccd0b1d5255d79a7ba32fd729a1db93d2f124/.github/workflows/scorecard.yml#L30

**Pull request closed without merging**